// @pnpm/logger/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.writeToConsole = exports.streamParser = exports.createStreamParser = exports.globalWarn = exports.globalInfo = exports.logger = void 0;
var logger_1 = require("./logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
Object.defineProperty(exports, "globalInfo", { enumerable: true, get: function () { return logger_1.globalInfo; } });
Object.defineProperty(exports, "globalWarn", { enumerable: true, get: function () { return logger_1.globalWarn; } });
var streamParser_1 = require("./streamParser");
Object.defineProperty(exports, "createStreamParser", { enumerable: true, get: function () { return streamParser_1.createStreamParser; } });
Object.defineProperty(exports, "streamParser", { enumerable: true, get: function () { return streamParser_1.streamParser; } });
var writeToConsole_1 = require("./writeToConsole");
Object.defineProperty(exports, "writeToConsole", { enumerable: true, get: function () { return writeToConsole_1.writeToConsole; } });
