// @glimmer/global-context/dist/modules/es5/index.js
/**
 * This package contains global context functions for Glimmer. These functions
 * are set by the embedding environment and must be set before initial render.
 *
 * These functions should meet the following criteria:
 *
 * - Must be provided by the embedder, due to having framework specific
 *   behaviors (e.g. interop with classic Ember behaviors that should not be
 *   upstreamed) or to being out of scope for the VM (e.g. scheduling a
 *   revalidation)
 * - Never differ between render roots
 * - Never change over time
 *
 */
import { DEBUG } from '@glimmer/env'; //////////

export var FEATURE_DEFAULT_HELPER_MANAGER = true; //////////

/**
 * Schedules a VM revalidation.
 *
 * Note: this has a default value so that tags can warm themselves when first loaded.
 */

export var scheduleRevalidate = function scheduleRevalidate() {};
/**
 * Schedules a destructor to run
 *
 * @param destroyable The destroyable being destroyed
 * @param destructor The destructor being scheduled
 */

export var scheduleDestroy;
/**
 * Finalizes destruction
 *
 * @param finalizer finalizer function
 */

export var scheduleDestroyed;
/**
 * Hook to provide iterators for `{{each}}` loops
 *
 * @param value The value to create an iterator for
 */

export var toIterator;
/**
 * Hook to specify truthiness within Glimmer templates
 *
 * @param value The value to convert to a boolean
 */

export var toBool;
/**
 * Hook for specifying how Glimmer should access properties in cases where it
 * needs to. For instance, accessing an object's values in templates.
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */

export var getProp;
/**
 * Hook for specifying how Glimmer should update props in cases where it needs
 * to. For instance, when updating a template reference (e.g. 2-way-binding)
 *
 * @param obj The object provided to get a value from
 * @param prop The prop to set the value at
 * @param value The value to set the value to
 */

export var setProp;
/**
 * Hook for specifying how Glimmer should access paths in cases where it needs
 * to. For instance, the `key` value of `{{each}}` loops.
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */

export var getPath;
/**
 * Hook for specifying how Glimmer should update paths in cases where it needs
 * to. For instance, when updating a template reference (e.g. 2-way-binding)
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */

export var setPath;
/**
 * Hook to warn if a style binding string or value was not marked as trusted
 * (e.g. HTMLSafe)
 */

export var warnIfStyleNotTrusted;
/**
 * Hook to customize assertion messages in the VM. Usages can be stripped out
 * by using the @glimmer/vm-babel-plugins package.
 */

export var assert;
/**
 * Hook to customize deprecation messages in the VM. Usages can be stripped out
 * by using the @glimmer/vm-babel-plugins package.
 */

export var deprecate;
var globalContextWasSet = false;
export default function setGlobalContext(context) {
  var _a;

  if (DEBUG) {
    if (globalContextWasSet) {
      throw new Error('Attempted to set the global context twice. This should only be set once.');
    }

    globalContextWasSet = true;
  }

  scheduleRevalidate = context.scheduleRevalidate;
  scheduleDestroy = context.scheduleDestroy;
  scheduleDestroyed = context.scheduleDestroyed;
  toIterator = context.toIterator;
  toBool = context.toBool;
  getProp = context.getProp;
  setProp = context.setProp;
  getPath = context.getPath;
  setPath = context.setPath;
  warnIfStyleNotTrusted = context.warnIfStyleNotTrusted;
  assert = context.assert;
  deprecate = context.deprecate;

  if (typeof ((_a = context.FEATURES) === null || _a === void 0 ? void 0 : _a.DEFAULT_HELPER_MANAGER) === 'boolean') {
    FEATURE_DEFAULT_HELPER_MANAGER = context.FEATURES.DEFAULT_HELPER_MANAGER;
  }
}
export var assertGlobalContextWasSet;
export var testOverrideGlobalContext;

if (DEBUG) {
  assertGlobalContextWasSet = function assertGlobalContextWasSet() {
    if (globalContextWasSet === false) {
      throw new Error('The global context for Glimmer VM was not set. You must set these global context functions to let Glimmer VM know how to accomplish certain operations. You can do this by importing `setGlobalContext` from `@glimmer/global-context`');
    }
  };

  testOverrideGlobalContext = function testOverrideGlobalContext(context) {
    var originalGlobalContext = globalContextWasSet ? {
      scheduleRevalidate: scheduleRevalidate,
      scheduleDestroy: scheduleDestroy,
      scheduleDestroyed: scheduleDestroyed,
      toIterator: toIterator,
      toBool: toBool,
      getProp: getProp,
      setProp: setProp,
      getPath: getPath,
      setPath: setPath,
      warnIfStyleNotTrusted: warnIfStyleNotTrusted,
      assert: assert,
      deprecate: deprecate
    } : null;

    if (context === null) {
      globalContextWasSet = false;
    } else {
      globalContextWasSet = true;
    } // We use `undefined as any` here to unset the values when resetting the
    // context at the end of a test.


    scheduleRevalidate = (context === null || context === void 0 ? void 0 : context.scheduleRevalidate) || undefined;
    scheduleDestroy = (context === null || context === void 0 ? void 0 : context.scheduleDestroy) || undefined;
    scheduleDestroyed = (context === null || context === void 0 ? void 0 : context.scheduleDestroyed) || undefined;
    toIterator = (context === null || context === void 0 ? void 0 : context.toIterator) || undefined;
    toBool = (context === null || context === void 0 ? void 0 : context.toBool) || undefined;
    getProp = (context === null || context === void 0 ? void 0 : context.getProp) || undefined;
    setProp = (context === null || context === void 0 ? void 0 : context.setProp) || undefined;
    getPath = (context === null || context === void 0 ? void 0 : context.getPath) || undefined;
    setPath = (context === null || context === void 0 ? void 0 : context.setPath) || undefined;
    warnIfStyleNotTrusted = (context === null || context === void 0 ? void 0 : context.warnIfStyleNotTrusted) || undefined;
    assert = (context === null || context === void 0 ? void 0 : context.assert) || undefined;
    deprecate = (context === null || context === void 0 ? void 0 : context.deprecate) || undefined;
    return originalGlobalContext;
  };
}
