// babel-plugin-ember-template-compilation/src/scope-locals.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ScopeLocals_mapping, _ScopeLocals_locals;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopeLocals = void 0;
/*
  This class exists because:
   - before template compilation starts, we need to pass a `locals` array to
     ember-template-compiler
   - the JSUtils API can mutate the scope during template compilation
   - those scope mutations need to update both the original `locals` array and
     our own name mapping, keeping them in sync.
*/
class ScopeLocals {
    constructor() {
        _ScopeLocals_mapping.set(this, {});
        _ScopeLocals_locals.set(this, []);
    }
    get locals() {
        return __classPrivateFieldGet(this, _ScopeLocals_locals, "f");
    }
    has(key) {
        return key in __classPrivateFieldGet(this, _ScopeLocals_mapping, "f");
    }
    get(key) {
        return __classPrivateFieldGet(this, _ScopeLocals_mapping, "f")[key];
    }
    isEmpty() {
        return __classPrivateFieldGet(this, _ScopeLocals_locals, "f").length === 0;
    }
    needsRemapping() {
        return Object.entries(__classPrivateFieldGet(this, _ScopeLocals_mapping, "f")).some(([k, v]) => k !== v);
    }
    entries() {
        return Object.entries(__classPrivateFieldGet(this, _ScopeLocals_mapping, "f"));
    }
    add(key, value) {
        __classPrivateFieldGet(this, _ScopeLocals_mapping, "f")[key] = value !== null && value !== void 0 ? value : key;
        if (!__classPrivateFieldGet(this, _ScopeLocals_locals, "f").includes(key)) {
            __classPrivateFieldGet(this, _ScopeLocals_locals, "f").push(key);
        }
    }
}
exports.ScopeLocals = ScopeLocals;
_ScopeLocals_mapping = new WeakMap(), _ScopeLocals_locals = new WeakMap();
