// @glimmer/global-context/dist/commonjs/es5/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = setGlobalContext;
exports.testOverrideGlobalContext = exports.assertGlobalContextWasSet = exports.deprecate = exports.assert = exports.warnIfStyleNotTrusted = exports.setPath = exports.getPath = exports.setProp = exports.getProp = exports.toBool = exports.toIterator = exports.scheduleDestroyed = exports.scheduleDestroy = exports.scheduleRevalidate = exports.FEATURE_DEFAULT_HELPER_MANAGER = void 0;

var _env = require("@glimmer/env");

/**
 * This package contains global context functions for Glimmer. These functions
 * are set by the embedding environment and must be set before initial render.
 *
 * These functions should meet the following criteria:
 *
 * - Must be provided by the embedder, due to having framework specific
 *   behaviors (e.g. interop with classic Ember behaviors that should not be
 *   upstreamed) or to being out of scope for the VM (e.g. scheduling a
 *   revalidation)
 * - Never differ between render roots
 * - Never change over time
 *
 */
//////////
var FEATURE_DEFAULT_HELPER_MANAGER = true; //////////

/**
 * Schedules a VM revalidation.
 *
 * Note: this has a default value so that tags can warm themselves when first loaded.
 */

exports.FEATURE_DEFAULT_HELPER_MANAGER = FEATURE_DEFAULT_HELPER_MANAGER;

var scheduleRevalidate = function scheduleRevalidate() {};
/**
 * Schedules a destructor to run
 *
 * @param destroyable The destroyable being destroyed
 * @param destructor The destructor being scheduled
 */


exports.scheduleRevalidate = scheduleRevalidate;
var scheduleDestroy;
/**
 * Finalizes destruction
 *
 * @param finalizer finalizer function
 */

exports.scheduleDestroy = scheduleDestroy;
var scheduleDestroyed;
/**
 * Hook to provide iterators for `{{each}}` loops
 *
 * @param value The value to create an iterator for
 */

exports.scheduleDestroyed = scheduleDestroyed;
var toIterator;
/**
 * Hook to specify truthiness within Glimmer templates
 *
 * @param value The value to convert to a boolean
 */

exports.toIterator = toIterator;
var toBool;
/**
 * Hook for specifying how Glimmer should access properties in cases where it
 * needs to. For instance, accessing an object's values in templates.
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */

exports.toBool = toBool;
var getProp;
/**
 * Hook for specifying how Glimmer should update props in cases where it needs
 * to. For instance, when updating a template reference (e.g. 2-way-binding)
 *
 * @param obj The object provided to get a value from
 * @param prop The prop to set the value at
 * @param value The value to set the value to
 */

exports.getProp = getProp;
var setProp;
/**
 * Hook for specifying how Glimmer should access paths in cases where it needs
 * to. For instance, the `key` value of `{{each}}` loops.
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */

exports.setProp = setProp;
var getPath;
/**
 * Hook for specifying how Glimmer should update paths in cases where it needs
 * to. For instance, when updating a template reference (e.g. 2-way-binding)
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */

exports.getPath = getPath;
var setPath;
/**
 * Hook to warn if a style binding string or value was not marked as trusted
 * (e.g. HTMLSafe)
 */

exports.setPath = setPath;
var warnIfStyleNotTrusted;
/**
 * Hook to customize assertion messages in the VM. Usages can be stripped out
 * by using the @glimmer/vm-babel-plugins package.
 */

exports.warnIfStyleNotTrusted = warnIfStyleNotTrusted;
var assert;
/**
 * Hook to customize deprecation messages in the VM. Usages can be stripped out
 * by using the @glimmer/vm-babel-plugins package.
 */

exports.assert = assert;
var deprecate;
exports.deprecate = deprecate;
var globalContextWasSet = false;

function setGlobalContext(context) {
  var _a;

  if (_env.DEBUG) {
    if (globalContextWasSet) {
      throw new Error('Attempted to set the global context twice. This should only be set once.');
    }

    globalContextWasSet = true;
  }

  exports.scheduleRevalidate = scheduleRevalidate = context.scheduleRevalidate;
  exports.scheduleDestroy = scheduleDestroy = context.scheduleDestroy;
  exports.scheduleDestroyed = scheduleDestroyed = context.scheduleDestroyed;
  exports.toIterator = toIterator = context.toIterator;
  exports.toBool = toBool = context.toBool;
  exports.getProp = getProp = context.getProp;
  exports.setProp = setProp = context.setProp;
  exports.getPath = getPath = context.getPath;
  exports.setPath = setPath = context.setPath;
  exports.warnIfStyleNotTrusted = warnIfStyleNotTrusted = context.warnIfStyleNotTrusted;
  exports.assert = assert = context.assert;
  exports.deprecate = deprecate = context.deprecate;

  if (typeof ((_a = context.FEATURES) === null || _a === void 0 ? void 0 : _a.DEFAULT_HELPER_MANAGER) === 'boolean') {
    exports.FEATURE_DEFAULT_HELPER_MANAGER = FEATURE_DEFAULT_HELPER_MANAGER = context.FEATURES.DEFAULT_HELPER_MANAGER;
  }
}

var assertGlobalContextWasSet;
exports.assertGlobalContextWasSet = assertGlobalContextWasSet;
var testOverrideGlobalContext;
exports.testOverrideGlobalContext = testOverrideGlobalContext;

if (_env.DEBUG) {
  exports.assertGlobalContextWasSet = assertGlobalContextWasSet = function assertGlobalContextWasSet() {
    if (globalContextWasSet === false) {
      throw new Error('The global context for Glimmer VM was not set. You must set these global context functions to let Glimmer VM know how to accomplish certain operations. You can do this by importing `setGlobalContext` from `@glimmer/global-context`');
    }
  };

  exports.testOverrideGlobalContext = testOverrideGlobalContext = function testOverrideGlobalContext(context) {
    var originalGlobalContext = globalContextWasSet ? {
      scheduleRevalidate: scheduleRevalidate,
      scheduleDestroy: scheduleDestroy,
      scheduleDestroyed: scheduleDestroyed,
      toIterator: toIterator,
      toBool: toBool,
      getProp: getProp,
      setProp: setProp,
      getPath: getPath,
      setPath: setPath,
      warnIfStyleNotTrusted: warnIfStyleNotTrusted,
      assert: assert,
      deprecate: deprecate
    } : null;

    if (context === null) {
      globalContextWasSet = false;
    } else {
      globalContextWasSet = true;
    } // We use `undefined as any` here to unset the values when resetting the
    // context at the end of a test.


    exports.scheduleRevalidate = scheduleRevalidate = (context === null || context === void 0 ? void 0 : context.scheduleRevalidate) || undefined;
    exports.scheduleDestroy = scheduleDestroy = (context === null || context === void 0 ? void 0 : context.scheduleDestroy) || undefined;
    exports.scheduleDestroyed = scheduleDestroyed = (context === null || context === void 0 ? void 0 : context.scheduleDestroyed) || undefined;
    exports.toIterator = toIterator = (context === null || context === void 0 ? void 0 : context.toIterator) || undefined;
    exports.toBool = toBool = (context === null || context === void 0 ? void 0 : context.toBool) || undefined;
    exports.getProp = getProp = (context === null || context === void 0 ? void 0 : context.getProp) || undefined;
    exports.setProp = setProp = (context === null || context === void 0 ? void 0 : context.setProp) || undefined;
    exports.getPath = getPath = (context === null || context === void 0 ? void 0 : context.getPath) || undefined;
    exports.setPath = setPath = (context === null || context === void 0 ? void 0 : context.setPath) || undefined;
    exports.warnIfStyleNotTrusted = warnIfStyleNotTrusted = (context === null || context === void 0 ? void 0 : context.warnIfStyleNotTrusted) || undefined;
    exports.assert = assert = (context === null || context === void 0 ? void 0 : context.assert) || undefined;
    exports.deprecate = deprecate = (context === null || context === void 0 ? void 0 : context.deprecate) || undefined;
    return originalGlobalContext;
  };
}
