// @pnpm/package-is-installable/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkPackage = exports.packageIsInstallable = exports.UnsupportedPlatformError = exports.UnsupportedEngineError = void 0;
const core_loggers_1 = require("@pnpm/core-loggers");
const checkEngine_1 = require("./checkEngine");
Object.defineProperty(exports, "UnsupportedEngineError", { enumerable: true, get: function () { return checkEngine_1.UnsupportedEngineError; } });
const checkPlatform_1 = require("./checkPlatform");
Object.defineProperty(exports, "UnsupportedPlatformError", { enumerable: true, get: function () { return checkPlatform_1.UnsupportedPlatformError; } });
const getSystemNodeVersion_1 = require("./getSystemNodeVersion");
function packageIsInstallable(pkgId, pkg, options) {
    const warn = checkPackage(pkgId, pkg, options);
    if (warn == null)
        return true;
    core_loggers_1.installCheckLogger.warn({
        message: warn.message,
        prefix: options.lockfileDir,
    });
    if (options.optional) {
        core_loggers_1.skippedOptionalDependencyLogger.debug({
            details: warn.toString(),
            package: {
                id: pkgId,
                name: pkg.name,
                version: pkg.version,
            },
            prefix: options.lockfileDir,
            reason: warn.code === 'ERR_PNPM_UNSUPPORTED_ENGINE' ? 'unsupported_engine' : 'unsupported_platform',
        });
        return false;
    }
    if (options.engineStrict)
        throw warn;
    return null;
}
exports.packageIsInstallable = packageIsInstallable;
function checkPackage(pkgId, manifest, options) {
    return (0, checkPlatform_1.checkPlatform)(pkgId, {
        cpu: manifest.cpu ?? ['any'],
        os: manifest.os ?? ['any'],
        libc: manifest.libc ?? ['any'],
    }) ?? ((manifest.engines == null)
        ? null
        : (0, checkEngine_1.checkEngine)(pkgId, manifest.engines, {
            node: options.nodeVersion ?? (0, getSystemNodeVersion_1.getSystemNodeVersion)(),
            pnpm: options.pnpmVersion,
        }));
}
exports.checkPackage = checkPackage;
