// @glimmer/global-context/dist/dev/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var env = require('@glimmer/env');

/**
 * This package contains global context functions for Glimmer. These functions
 * are set by the embedding environment and must be set before initial render.
 *
 * These functions should meet the following criteria:
 *
 * - Must be provided by the embedder, due to having framework specific
 *   behaviors (e.g. interop with classic Ember behaviors that should not be
 *   upstreamed) or to being out of scope for the VM (e.g. scheduling a
 *   revalidation)
 * - Never differ between render roots
 * - Never change over time
 *
 */

//////////

/**
 * Interfaces
 *
 * TODO: Move these into @glimmer/interfaces, move @glimmer/interfaces to
 * @glimmer/internal-interfaces.
 */
//////////
/**
 * Schedules a VM revalidation.
 *
 * Note: this has a default value so that tags can warm themselves when first loaded.
 */
exports.scheduleRevalidate = () => {};

/**
 * Schedules a destructor to run
 *
 * @param destroyable The destroyable being destroyed
 * @param destructor The destructor being scheduled
 */
exports.scheduleDestroy = void 0;

/**
 * Finalizes destruction
 *
 * @param finalizer finalizer function
 */
exports.scheduleDestroyed = void 0;

/**
 * Hook to provide iterators for `{{each}}` loops
 *
 * @param value The value to create an iterator for
 */
exports.toIterator = void 0;

/**
 * Hook to specify truthiness within Glimmer templates
 *
 * @param value The value to convert to a boolean
 */
exports.toBool = void 0;

/**
 * Hook for specifying how Glimmer should access properties in cases where it
 * needs to. For instance, accessing an object's values in templates.
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */
exports.getProp = void 0;

/**
 * Hook for specifying how Glimmer should update props in cases where it needs
 * to. For instance, when updating a template reference (e.g. 2-way-binding)
 *
 * @param obj The object provided to get a value from
 * @param prop The prop to set the value at
 * @param value The value to set the value to
 */
exports.setProp = void 0;

/**
 * Hook for specifying how Glimmer should access paths in cases where it needs
 * to. For instance, the `key` value of `{{each}}` loops.
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */
exports.getPath = void 0;

/**
 * Hook for specifying how Glimmer should update paths in cases where it needs
 * to. For instance, when updating a template reference (e.g. 2-way-binding)
 *
 * @param obj The object provided to get a value from
 * @param path The path to get the value from
 */
exports.setPath = void 0;

/**
 * Hook to warn if a style binding string or value was not marked as trusted
 * (e.g. HTMLSafe)
 */
exports.warnIfStyleNotTrusted = void 0;

/**
 * Hook to customize assertion messages in the VM. Usages can be stripped out
 * by using the @glimmer/vm-babel-plugins package.
 */
exports.assert = void 0;

/**
 * Hook to customize deprecation messages in the VM. Usages can be stripped out
 * by using the @glimmer/vm-babel-plugins package.
 */
exports.deprecate = void 0;

//////////

let globalContextWasSet = false;
function setGlobalContext(context) {
  if (env.DEBUG) {
    if (globalContextWasSet) {
      throw new Error('Attempted to set the global context twice. This should only be set once.');
    }
    globalContextWasSet = true;
  }
  exports.scheduleRevalidate = context.scheduleRevalidate;
  exports.scheduleDestroy = context.scheduleDestroy;
  exports.scheduleDestroyed = context.scheduleDestroyed;
  exports.toIterator = context.toIterator;
  exports.toBool = context.toBool;
  exports.getProp = context.getProp;
  exports.setProp = context.setProp;
  exports.getPath = context.getPath;
  exports.setPath = context.setPath;
  exports.warnIfStyleNotTrusted = context.warnIfStyleNotTrusted;
  exports.assert = context.assert;
  exports.deprecate = context.deprecate;
}
exports.assertGlobalContextWasSet = void 0;
exports.testOverrideGlobalContext = void 0;
if (env.DEBUG) {
  exports.assertGlobalContextWasSet = () => {
    if (globalContextWasSet === false) {
      throw new Error('The global context for Glimmer VM was not set. You must set these global context functions to let Glimmer VM know how to accomplish certain operations. You can do this by importing `setGlobalContext` from `@glimmer/global-context`');
    }
  };
  exports.testOverrideGlobalContext = context => {
    let originalGlobalContext = globalContextWasSet ? {
      scheduleRevalidate: exports.scheduleRevalidate,
      scheduleDestroy: exports.scheduleDestroy,
      scheduleDestroyed: exports.scheduleDestroyed,
      toIterator: exports.toIterator,
      toBool: exports.toBool,
      getProp: exports.getProp,
      setProp: exports.setProp,
      getPath: exports.getPath,
      setPath: exports.setPath,
      warnIfStyleNotTrusted: exports.warnIfStyleNotTrusted,
      assert: exports.assert,
      deprecate: exports.deprecate
    } : null;
    if (context === null) {
      globalContextWasSet = false;
    } else {
      globalContextWasSet = true;
    }

    // We use `undefined as any` here to unset the values when resetting the
    // context at the end of a test.
    exports.scheduleRevalidate = context?.scheduleRevalidate || undefined;
    exports.scheduleDestroy = context?.scheduleDestroy || undefined;
    exports.scheduleDestroyed = context?.scheduleDestroyed || undefined;
    exports.toIterator = context?.toIterator || undefined;
    exports.toBool = context?.toBool || undefined;
    exports.getProp = context?.getProp || undefined;
    exports.setProp = context?.setProp || undefined;
    exports.getPath = context?.getPath || undefined;
    exports.setPath = context?.setPath || undefined;
    exports.warnIfStyleNotTrusted = context?.warnIfStyleNotTrusted || undefined;
    exports.assert = context?.assert || undefined;
    exports.deprecate = context?.deprecate || undefined;
    return originalGlobalContext;
  };
}

exports.default = setGlobalContext;
