// ember-tracked-storage-polyfill/commonjs/index.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setValue = exports.getValue = exports.createStorage = void 0;
const tracking_1 = require("@glimmer/tracking");
const debug_1 = require("@ember/debug");
class TrackedStorageImpl {
    constructor(initialValue, isEqual) {
        this._value = this._lastValue = initialValue;
        this._isEqual = isEqual;
    }
}
__decorate([
    tracking_1.tracked
], TrackedStorageImpl.prototype, "_value", void 0);
function tripleEq(a, b) {
    return a === b;
}
function createStorage(initialValue, isEqual = tripleEq) {
    debug_1.assert('the second parameter to `createStorage` must be an equality function or undefined', typeof isEqual === 'function');
    return new TrackedStorageImpl(initialValue, isEqual);
}
exports.createStorage = createStorage;
function getValue(storage) {
    debug_1.assert('getValue must be passed a tracked store created with `createStorage`.', storage instanceof TrackedStorageImpl);
    return storage._value;
}
exports.getValue = getValue;
function setValue(storage, value) {
    debug_1.assert('setValue must be passed a tracked store created with `createStorage`.', storage instanceof TrackedStorageImpl);
    const { _isEqual: isEqual, _lastValue: lastValue } = storage;
    if (!isEqual(value, lastValue)) {
        storage._value = storage._lastValue = value;
    }
}
exports.setValue = setValue;
